src/azul/terra.py

```diff
@@ -352,8 +439,48 @@ def _repository_endpoint(self, *path: str) -> str:
         return furl(config.tdr_service_url,
                     path=('api', 'repository', 'v1', *path)).url
 
+    def _check_response(self,
+                        endpoint: str,
+                        response: urllib3.HTTPResponse
+                        ) -> MutableJSON:
+        if response.status == 200:
+            return json.loads(response.data)
+        elif response.status == 401:
+            raise self._insufficient_access(endpoint)
+        else:
+            raise RequirementError('Unexpected response from TDR API', response.status)
+
+    def snapshot_ids_by_name(self) -> Dict[str, str]:
+        """
+        List the TDR snapshots accessible to the current credentials.
+        """
+        limit = 100
+        endpoint = self._repository_endpoint('snapshots')
+        response = self._request('GET', endpoint, fields=dict(limit=str(limit)))
+        response = self._check_response(endpoint, response)
+        # FIXME Implement paging
+        #       https://github.com/DataBiosphere/azul/issues/3093
+        require(response['total'] <= limit, 'Too many snapshots', response['total'])
+        return {
+            snapshot['name']: snapshot['id']
+            for snapshot in response['items']
+        }
+
+    @classmethod
+    def with_service_account_credentials(cls) -> 'TDRClient':
+        return cls(credentials_provider=ServiceAccountCredentialsProvider())
+
+    @classmethod
+    def with_user_credentials(cls, token: OAuth2) -> 'TDRClient':
+        return cls(credentials_provider=UserCredentialsProvider(token))
+
 
-class TerraDRSClient(DRSClient, TerraClient):
+@attr.s(auto_attribs=True, kw_only=True, frozen=True)
+class TerraDRSClient(DRSClient, TDRClient):
+    credentials_provider: CredentialsProvider = ServiceAccountCredentialsProvider()
```

Comment:
Given that ServiceAccountCredentialsProvider isn't stateless and that its state is user-specific, I don't think it's a good idea to initialize this at the class level. It might even be a gaping security hole.